### PR TITLE
Fix Environment.UserInteractive on Windows

### LINKED ISSUE #770: Should we revise implementation of Environment.UserInteractive?
This is currently implemented as `return true;` but since in 3.0 we're adding UI stacks to .NET Core, maybe we should implement this since this is usually used before popping a UI?

The Windows version of the algorithm is copied all over the place (touching it in dotnet/corefx#39285, dotnet/winforms#1247).

### PR DESCRIPTION
**Pull request by danmoseley**
Fix https://github.com/dotnet/runtime/issues/770

### CODE REVIEW

File: src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs
Comment (jkotas) [inline src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs]: Microsoft.Win32.SystemEvents is a NuGet package that targets netstandard2.0. Changing this to use `Environment.UserInteractive` is going to break this NuGet package on .NET Core 2.x and 3.x.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]: If you open System.Private.CoreLib.csproj in Visual Studio, are these files going to show up correctly in the project view? (ie in reasonable place of the hiearchy)

I know that we had troubles with this in the past and kept adding the `<Link>...</Link>` elements as workaround.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Please do not use `Marshal.SizeOf`. It is slow expensive API. Use `sizeof` instead.

See https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/interop-guidelines.md#definition

File: src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs
Comment (jkotas) [inline src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs]: You can delete this condition if you would like to. We are not supporting Win95 anymore.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: I think that Nano should be rather treated as Unix, and always return true.

I would delete this special early out.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Note that the improper use of this API for detecting whether the process is running as Windows Service is not going to work on Nano in any case.

File: src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs
Comment (danmosemsft) [reply src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs]: oh! - of course

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]: VS doesn't even show me any files originating from this .projitems file that are conditionalized like `'$(TargetsWindows)' == 'true'`. It does show some files that are behind other conditions.

We need to figure out what criteria it uses. Long time ago, when I worked on this, the critera was something like -- all files in the main project, whatever their condition; no files in an imported project; reverse either case with metadata `<Visible>false</Visible>` (which could be applied by default)

I will take a look at this separately after your change, if you don't solve it already.

File: src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs
Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: OK. Why might I want to ever use `Marshal.SizeOf` in this context - and if there are no reasons, perhaps we should have an analyzer?

I read through https://github.com/dotnet/csharplang/issues/1828 in the hope it would explain this, but it only suggests why `Marshal.SizeOf` is not always functionally equivalent.

File: src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs
Comment (jkotas) [inline src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs]: Is this test going to work on Windows Nano?

File: src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs
Comment (jkotas) [inline src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs]:
```suggestion
                IntPtr hwinsta = Interop.User32.GetProcessWindowStation();
```

File: src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs
Comment (jkotas) [inline src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs]: The PInvoke definition of GetUserObjectInformation is not quite right:
- The 3rd argument should better be `void*` so that it can be used to any type of information and not just this specific one (`void*` has smaller marshaling overhead as well)
- The 4th and 5th arguments should be `uint`

File: src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs
Comment (danmosemsft) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs]: Good point - I still had it backwards in my head. OK, I gave it another shot. I can't get these tests to pass currently anyway.
https://github.com/dotnet/runtime/pull/1234

File: src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs
Comment (danmosemsft) [reply src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs]: ah yes, I didn't see it was PVOID and that is called out. https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/interop-pinvokes.md

File: src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Yes, the analyzer for this is listed in https://github.com/dotnet/corefx/issues/40739.

The only situation where `Marshal.SizeOf`, `Marshal.PtrToStructure` and friends make sense are non-blittable types. We try hard to avoid non-blittable structs for interop.

**Comment by jkotas** [comment]
Build break..

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]: > VS doesn't even show me any files originating from this .projitems fil

It does show it to me (you may need to open System.Private.CoreLib.sln):

image

I see that the VS has troubles with just these new files added against `$(LibrariesProjectRoot)` without `<Link>...</Link>`.

File: src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Nit: This file is not in the common yet...

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]:
```suggestion
    <Compile Include=""$(CommonPath)Interop\Windows\User32\Interop.GetProcessWindowStation.cs">
```

**Comment by danmosemsft** [comment]
All failures are from the bad ARM legs.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]:
```suggestion
    <Compile Include=""$(CommonPath)Interop\Windows\User32\Interop.GetUserObjectInformation.cs">
```

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]:
```suggestion
    <Compile Include=""$(CommonPath)Interop\Windows\User32\Interop.USEROBJECTFLAGS.cs">
```

File: src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Nit: The `(void*)` cast is unnecessary

**Comment by danmosemsft** [comment]
Any more feedback? I can't merge without dismissing your review -- not sure of the protocol

CI is fine, except for an unrelated bug I logged.

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs
Comment (ericstj) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: It looks like you forgot to invert this logic.   Also, should we add back the static property like we had in desktop?
https://referencesource.microsoft.com/#mscorlib/system/environment.cs,1435

Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Ack - thank you @ericstj !

Re the static, @jkotas pointed out that there may be a way to change a process to/from interactive - although I didn't do it successfully myself. See mention here https://github.com/dotnet/runtime/issues/770#issuecomment-570006411 I saw no reason to cache the value.

Comment (ericstj) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Ok, I'll make the simple fix for now.  FWIW I'm now able to reproduce the test failure, so I can confirm fail/then-fix.

Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: That would explain https://github.com/dotnet/runtime/issues/1724 also. I was not albe to run the service controller tests locally because of #1262. I tested locally by hand instead, and apparently did so before I made this mistake.

Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs]: Comments crossed. Great, I was about to do it, but apparently you've already got those tests working so you can verify it as well. thanks

**Pull request merged**
